File: boj_codes/C++17/12865.cpp
// https://www.acmicpc.net/problem/12865
// First Written : 20210120
// Last Modified : 20210120

#include <cstdio>
#define  MAX(X,Y) (((X) > (Y)) ? (X) : (Y))

int main(){
    int N, max_load;
    int cur_weight, cur_value;
    scanf("%d %d", &N, &max_load);

    int weights[N+1] = {0};
    int values[N+1] = {0};

    int dp[N + 1][max_load + 1];

    for(int i = 1; i <= N; i++){
        scanf("%d %d", &weights[i], &values[i]);
    }


    for(int idx = 0; idx <= N; idx++){
        cur_weight = weights[idx];
        cur_value = values[idx];

        for(int load = 0; load <= max_load; load++){
            if(idx == 0 || load == 0){
                dp[idx][load] = 0;
                continue;
            }
            
            else if(cur_weight > load){
                dp[idx][load] = dp[idx-1][load];
            }
            else{
                dp[idx][load] = MAX(
                    dp[idx-1][load],
                    dp[idx-1][load-cur_weight] + cur_value);
            }
        }
    }
    
    printf("%d", dp[N][max_load]);
    return 0;
}

File: old_contents/boj_codes/Python3/12865.py
# https://www.acmicpc.net/problem/12865
# 20201126

import sys

if __name__ == "__main__":
    N, max_load = map(int, sys.stdin.readline().split())

    dp = [[None] * (max_load + 1) for _ in range(N + 1)]

    weights = [0]
    values = [0]

    for _ in range(N):
        weight, value = map(int, sys.stdin.readline().split())

        weights.append(weight)
        values.append(value)

    for idx in range(N + 1):
        cur_weight = weights[idx]
        cur_value = values[idx]

        for load in range(max_load + 1):
            if idx == 0 or load == 0:
                dp[idx][load] = 0
                continue

            elif cur_weight > load:
                dp[idx][load] = dp[idx - 1][load]

            else:
                dp[idx][load] = max(dp[idx - 1][load],
                                    dp[idx - 1][load - cur_weight] + cur_value)


    print(dp[N][max_load])

